R/slice.R

```diff
@@ -444,24 +444,20 @@ check_bool <- function(x, arg = caller_arg(x), call = caller_env()) {
   }
 }
 
-get_slice_size <- function(n, prop, allow_negative = TRUE, error_call = caller_env()) {
+get_slice_size <- function(n, prop, error_call = caller_env()) {
   slice_input <- check_slice_n_prop(n, prop, error_call = error_call)
 
   if (slice_input$type == "n") {
     if (slice_input$n >= 0) {
       function(n) floor(slice_input$n)
-    } else if (allow_negative) {
-      function(n) ceiling(n + slice_input$n)
     } else {
-      abort("`n` must be positive.", call = error_call)
+      function(n) ceiling(n + slice_input$n)
     }
   } else if (slice_input$type == "prop") {
     if (slice_input$prop >= 0) {
       function(n) floor(slice_input$prop * n)
-    } else if (allow_negative) {
-      function(n) ceiling(n + slice_input$prop * n)
     } else {
-      abort("`prop` must be positive.", call = error_call)
+      function(n) ceiling(n + slice_input$prop * n)
```

Comment:
```suggestion
      function(n) max(ceiling(n + slice_input$prop * n), 0L)
```

I think you could "fix" all of the problems by creating an invariant that `get_slice_size()` always returns a positive size. You'd apply this trick to `ceiling(n + slice_input$n)` too. I think that `slice_min/max/head/tail()` are already working right with OOB negative sizes, but this would be a centralized guarantee that they wouldn't have to even worry about it.

Then I think `sample_int()` is really just this, because it should handle the zero size case already

```r
sample_int <- function(n, size, replace = FALSE, wt = NULL, call = caller_env()) {
  if (!replace && n < size) {
    size <- n
  }
  sample.int(n, size, prob = wt, replace = replace)
}
```

Reply:
In theory you'd have a symmetrical `min(floor(slice_input$prop * n), n)` clamp on the positive branches, which would allow you to remove this kind of check in `slice_head()` and `slice_tail()`

```r
if (to > n) {
  to <- n
}
```

and this in `sample_int()`

```r
if (!replace && n < size) {
  size <- n
}
```

You'd just have 1 special case for the `slice_sample(replace = TRUE)` case to _not_ clamp the positive branch (because, i.e., `slice_sample(mtcars, n = 100, replace = TRUE)` makes sense)

Reply:
Maybe this should be a separate PR, but I do think it would dramatically simplify things

Reply:
Yeah, I think that's right, now that I understand what negative values actually do.